app/models/spreadsheet_entry.rb

```diff
@@ -4,31 +4,93 @@ class SpreadsheetEntry
-                :display_as, :category, :primary_taxon_id, :price, :on_hand, :on_demand,
-                :tax_category_id, :shipping_category_id, :description
+                :display_as, :category, :primary_taxon_id, :price, :on_hand, :count_on_hand, :on_demand,
+                :tax_category_id, :shipping_category_id, :description, :import_date
 
   def initialize(attrs)
-    @product_validations = {}
+    #@product_validations = {}
+    @validates_as = ''
+
+    #validate_custom_unit_fields(attrs, is_inventory)
+    convert_custom_unit_fields(attrs)
 
     attrs.each do |k, v|
       if self.respond_to?("#{k}=")
         send("#{k}=", v) unless non_product_attributes.include?(k)
       else
-        # Trying to assign unknown attribute. Record this and give feedback or just ignore silently?
+        # Trying to assign unknown attribute... record this and give feedback or
+        # just continue to ignore silently?
       end
     end
   end
 
+  def unit_scales
+    {
+      'g' => {scale: 1, unit: 'weight'},
+      'kg' => {scale: 1000, unit: 'weight'},
+      't' => {scale: 1000000, unit: 'weight'},
+      'ml' => {scale: 0.001, unit: 'volume'},
+      'l' => {scale: 1, unit: 'volume'},
+      'kl' => {scale: 1000, unit: 'volume'}
+    }
+  end
+
+  def convert_custom_unit_fields(attrs)
+    # unit unit_type variant_unit_name   ->    unit_value  variant_unit_scale   variant_unit
+    # 250    ml       nil      ....              0.25        0.001               volume
+    # 50     g        nil      ....              50          1                   weight
+    # 2     kg        nil      ....              2000        1000                weight
+    # 1     nil       bunches  ....              1           null                items
+
+    attrs['variant_unit'] = nil
+    attrs['variant_unit_scale'] = nil
+    attrs['unit_value'] = nil
+
+    if attrs.key?('units') && attrs['units'].present?
+      attrs['unscaled_units'] = attrs['units']
+    end
+
+    if attrs.key?('units') && attrs.key?('unit_type') && attrs['units'].present? && attrs['unit_type'].present?
+      units = attrs['units'].to_f
+      unit_type = attrs['unit_type'].to_s.downcase
+
+      if valid_unit_type? unit_type
+        attrs['variant_unit'] = unit_scales[unit_type][:unit]
+        attrs['variant_unit_scale'] = unit_scales[unit_type][:scale]
+        attrs['unit_value'] = (units || 0) * attrs['variant_unit_scale']
+      end
+    end
+
+    return unless attrs.key?('units') && attrs.key?('variant_unit_name') && attrs['units'].present? && attrs['variant_unit_name'].present?
+
+    attrs['variant_unit'] = 'items'
+    attrs['variant_unit_scale'] = nil
+    attrs['unit_value'] = units || 1
+  end
+
   def persisted?
     false #ActiveModel
   end
 
+  def is_a_valid?(type)
+    #@validates_as[type]
+    @validates_as == type
+  end
+
+  def is_a_valid(type)
+    #@validates_as.push type
+    @validates_as = type
+  end
```

Comment:
watch out, these methods might have confusing names.

Reply:
They produce some nice lines, like: 
`save_new_product(entry) if entry.is_a_valid? 'new_product'`
What do you think?